Use physical path instead of cached GOROOT in function getGoroot (fixes #433, #1658)

ardnew opened the pull request:
I keep running into the same error as others (#433 and #1658), where TinyGo resolves an incorrect Go runtime version, and generates the message:

```
error: requires go version 1.11 through 1.15, got go0.1
error: command failed
```

On Linux, this appears to be caused by the symlink created in the cached `GOROOT`:

```sh
andrew@trip ~ readlink -f $( tinygo info teensy40 | \grep -oP "GOROOT:\s*\K.+" )/src/runtime/internal/sys/zversion.go
/usr/local/src/tinygo/src/runtime/internal/sys/zversion.go
andrew@trip ~ cat /usr/local/src/tinygo/src/runtime/internal/sys/zversion.go
package sys

const TheVersion = `go0.1.0`
```

The problem is manifested here: https://github.com/tinygo-org/tinygo/blob/c60c36f0a8bbfd62fc61a583d4d2c7a8bf64bd58/goenv/version.go#L51-L52

This PR adds logic that tries to resolve a better path to `zversion.go` by checking if `GOROOT/bin` is a symlink (which it always will be inside TinyGo's cached `GOROOT`), and using the real, physical path to `GOROOT/bin` as the root directory for finding `src/runtime/internal/sys/zversion.go`. This way we are using the actual Go runtime directory's `zversion.go`, and not the one stored in TinyGo's sources.

Note that if this method fails, it falls back to using the original method currently implemented.

aykevl commented:
I would argue the issue here is not in `GorootVersionString` but in how GOROOT is determined in the first place. A GOROOT set to the cached GOROOT (in ~/.cache/tinygo) will surely cause trouble. Therefore, I think the more straightforward solution here is to ignore the GOROOT environment variable if it points into the cache directory.

When writing this logic I had not considered that GOROOT could be set to the cached GOROOT of TinyGo. But now I see this, it does make sense to do so in an IDE (which might pass on this environment variable to tools it invokes).

ardnew commented:
> Therefore, I think the more straightforward solution here is to ignore the GOROOT environment variable if it points into the cache directory.

Without a properly-defined environment variable, how would you suggest a TinyGo GOROOT be resolved? Perhaps this is an unrecoverable error?

> A GOROOT set to the cached GOROOT (in ~/.cache/tinygo) will surely cause trouble.

I was under the impression that a GOROOT pointing to one of the cached directories (`~/.cache/tinygo/goroot-goVER-HASH-syscall`) was the _only_ permissible GOROOT definition for TinyGo - hence, this is why we use `tinygo info target` to define GOROOT when launching IDEs and tools. After all, these cached TinyGo GOROOTs are very different (that is, they have different content) from any installed standard Go's GOROOT.

aykevl commented:
There are three different GOROOT-like paths in TinyGo:

* The path which points to the root of the Go toolchain (GOROOT).
* The path which points to the root of the TinyGo toolchain (TINYGOROOT).
* The path which merges the two and is stored in the TinyGo cache (cached GOROOT).

TinyGo will try various things to obtain a valid GOROOT, first trying the GOROOT environment variable, then trying the location of the `go` command, and finally trying a few fallback paths. For details, see:

https://github.com/tinygo-org/tinygo/blob/138befd8a25d2a5c65e2e740038a8c3a44e95222/goenv/goenv.go#L142-L144

TinyGo will then merge the GOROOT and TINYGOROOT to a cached GOROOT, and will pass this path on to the `go list` command to resolve package dependencies. This cached GOROOT is also used in IDEs so that IDEs will resolve the correct dependencies (the machine package, for example).

The bug here is that this cached GOROOT is then set as the GOROOT environment variable when calling TinyGo. TinyGo expects the root of the Go toolchain but instead gets the cached GOROOT, which it then tries to merge _again_ with TINYGOROOT. Which of course gives a rather odd result.

I think the best way to fix this is by improving the logic of `getGoroot` to either detect this and ignore the passed in GOROOT (simple fix, perhaps not entirely correct) or maybe better to return the root of the Go installation by reading it from this cached GOROOT.

I hope that clears things up!

ardnew committed:
check for cached GOROOT at function getGoroot

ardnew committed:
use legacy package "os" instead of "io/fs" to provide type FileMode

ardnew committed:
use legacy ioutil.ReadDir instead of os.ReadDir

ardnew committed:
use legacy os.ModeType bitmask instead of os.FileMode method Type

ardnew commented:
> I think the best way to fix this is by improving the logic of getGoroot to either detect this and ignore the passed in GOROOT (simple fix, perhaps not entirely correct) or maybe better to return the root of the Go installation by reading it from this cached GOROOT.

@aykevl I've updated this PR to implement your suggestion here (reverted my original commit). So instead of modifying `func GorootVersionString`, I've modified `getGoroot` as follows:

#### Previous logic
```
function getGoroot:
  if environment variable ${GOROOT} is defined:
    return ${GOROOT}
  ...
```

#### Changed logic
```
function getGoroot:
  if environment variable ${GOROOT} is defined:
    if the path at ${GOROOT} appears to be a TinyGo cache:
      return the standard Go GOROOT referenced by the TinyGo cache ${GOROOT}
    return ${GOROOT}
  ...
```

To identify a GOROOT as a TinyGo cache, it needs to meet the following conditions:
1. It is a directory containing at least the following files with exactly these file modes:
   - `src`: `ModeDir`
   - `bin`: `ModeSymlink`
   - `lib`: `ModeSymlink`
   - `pkg`: `ModeSymlink`
2. All of the above symlinks point to subdirectories of the _**exact same parent directory**_ (the GOROOT)

This resolved the issue for me locally, but I definitely need some help verifying these conditions are correct in general (not just Linux).

ardnew commented:
Perhaps we can place a file in the root of the TinyGo cache upon cache creation, or even in its parent directory (e.g., `~/.cache/tinygo/`), whose existence means that path is a cached GOROOT?

Or maybe we can reliably identify the path to where these caches are being created, and assume any subdirectory of that path is a cached GOROOT?

aykevl commented:
Yes, that code looks roughly correct. But I think it can be simplified a lot. Because TinyGo itself creates this cached GOROOT, we don't need to guess what's in there. So I think the following would be good enough:

```diff
 	goroot := os.Getenv("GOROOT")
 	if goroot != "" {
+		if symlink, err := os.Readlink(filepath.Join(goroot, "bin")); err == nil {
+			return filepath.Base(symlink)
+		}
 		return goroot
 	}
```

I think it's unlikely upstream Go will ever make the bin directory a symlink.

ardnew commented:
OK, so no need to really inspect the `GOROOT` that rigorously, I'll update the PR

(Also I believe you meant `filepath.Dir` instead of `filepath.Base`)

ardnew committed:
goenv: simplify checks for TinyGo cache in func getGoroot()

ardnew commented:
@aykevl Updated once more with your latest comments. Please review.

ardnew commented:
Been testing this change with all sorts of valid and invalid configurations:

- `GOROOT` environment variable
   - Using `tinygo` command manually 
- VSCode with TinyGo extension
   - Toggling between different targets
   - Setting `{"go.toolsEnvVars":{"GOROOT"}}` in `.vscode/settings.json` manually
- SublimeText3 with GoSublime, gopls, and goguru extensions
   - Setting `GOPATH` in margo configuration
   - Setting `GS_GOPATH` and other `GS_*` environment variables

Not once have I encountered the error message `got go0.1` again! I think this is a stable change.

aykevl commented:
Thank you! Looks good to me.

There was one failing test but that looks like a flaky test, unrelated to this PR.
